client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,576 @@
+				trace!(
+					target: "peerset",
+					"Disconnecting previously reserved node {} on {:?}.",
+					node.key(), self.set_id
+				);
+			} else {
+				// Otherwise, just count occupied slots for the node.
+				match node.get() {
+					ConnectionState::In => self.num_in += 1,
+					ConnectionState::Out => self.num_out += 1,
+					ConnectionState::NotConnected => {},
+				}
+			}
+		} else {
+			error!(
+				target: "peerset",
+				"Invalid state: reserved node {} not in the list of known nodes.",
+				peer_id,
+			);
+		}
+	}
+
+	fn on_set_reserved_peers(&mut self, peer_ids: HashSet<PeerId>) {
+		// Determine the difference between the current group and the new list.
+		let to_insert = peer_ids.difference(&self.reserved_nodes).cloned().collect::<Vec<_>>();
+		let to_remove = self.reserved_nodes.difference(&peer_ids).cloned().collect::<Vec<_>>();
+
+		for node in to_insert {
+			self.on_add_reserved_peer(node);
+		}
+
+		for node in to_remove {
+			self.on_remove_reserved_peer(node);
+		}
+	}
+
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
+		self.reserved_only = reserved_only;
+
+		if reserved_only {
+			// Disconnect all non-reserved peers.
+			for peer_id in self.connected_peers().cloned().collect::<Vec<_>>().iter() {
+				self.peer(&peer_id)
+					.into_connected()
+					.expect(
+						"We are enumerating connected peers, therefore the peer is connected; qed",
+					)
+					.disconnect();
+			}
+		} else {
+			// Try connecting to non-reserved peers.
+			self.alloc_slots();
+		}
+	}
+
+	fn on_add_peer(&mut self, peer_id: PeerId) {
+		if self.nodes.insert(peer_id, ConnectionState::NotConnected).is_none() {
+			self.alloc_slots();
+		}
+	}
+
+	fn on_remove_peer(&mut self, peer_id: PeerId) {
+		// Don't do anything if the node is reserved.
+		if self.reserved_nodes.contains(&peer_id) {
+			return
+		}
+
+		self.peer(&peer_id).into_connected().map(|connected_peer| connected_peer.disconnect());
+		self.nodes.remove(&peer_id);
+	}
+
+	fn on_incoming_connection(&mut self, peer_id: PeerId, incoming_index: IncomingIndex) {
+		if self.reserved_only && !self.reserved_nodes.contains(&peer_id) {
+			let _ = self.to_notifications.unbounded_send(Message::Reject(incoming_index));
+			return
+		}
+
+		let not_connected = match self.peer(&peer_id) {
+			// If we're already connected, don't answer, as the docs mention.
+			Peer::Connected(_) => return,
+			Peer::NotConnected(peer) => peer,
+			// Adding incoming peer to our set of peers even before we decide whether to accept
+			// it is questionable, but this is how it was implemented in the original `Peerset`.
+			Peer::Unknown(peer) => peer.discover(),
+		};
+
+		if not_connected.is_banned() {
+			let _ = self.to_notifications.unbounded_send(Message::Reject(incoming_index));
+			return;
+		}
+
+		let _ = not_connected.try_accept_incoming(incoming_index);
+	}
+
+	fn peer<'a>(&'a mut self, peer_id: &'a PeerId) -> Peer<'a> {
```

Comment:
I'm wondering do we actually need this functionality at all? What if the `peers` in `ProtocolController` would store a `PeerState` and `on_incoming_connection()` would check that state and execute code based on the state instead of crafting this temporary `Peer` object. 

```rust
enum Direction {
    Inbound,
    Outbound,
}

enum PeerState {
    Connected(Direction),
    NotConnected,
    Unknown,
}
```

Unless I'm glancing over some other issues, this would simplify the implementation as we could get rid of these member functions from different flavors of `Peer` and just update the `peers` hashmap and the peer state based on what action is being performed. This would also allow us to keep the `Unknown` state for now.

Reply:
To elaborate, `peers` in `ProtocolController` would now be `HashMap<PeerId, PeerState>` and when, for example `alloc_slots()` or `on_incoming_connection()` is called, we would perform the necessary operation based on the state without creating this temporary `Peer` object. The code can still be reused for the most part but just moved to a different place and not have, for example `disconnect()`, be a member function of `ConnectedPeer`.

Reply:
Thanks for the proposal, I'll give it a try.

Reply:
> This would also allow us to keep the `Unknown` state for now.

Actually, this `Unknown` state is purely an abstraction over a peer that is not present in the `HashMap`, which can be promoted to `NotConnected` peer (which is present in the `HashMap`). So I don't think it makes sense to put these `Unknown` records directly into the `HashMap`.

Reply:
> Did a first pass. There are some problems we need to address, especially regarding `Peer`. I hope we can get rid of it and just `match` on peer state and then perform the correct action.

I've removed the `Peer` abstraction completely. This lead to the need of keeping track of connection states and sending messages to `Notifications` manually in all the member functions, but overall the level of complexity seems more or less acceptable.